File: core/src/com/alta_v2/physics/task/moveNpc/MoveNpcTask.java
package com.alta_v2.physics.task.moveNpc;

import com.alta_v2.physics.executionContext.AltitudeMap;
import com.alta_v2.physics.executionContext.Tenant;
import com.alta_v2.physics.executionContext.reserveData.ReservableActor;
import com.alta_v2.physics.executionContext.reserveData.ReservablePoint;
import com.alta_v2.physics.task.MovementDirection;
import com.alta_v2.physics.task.ResultTiledMapTask;
import com.alta_v2.physics.task.resultObserver.BaseTaskResultObserver;
import com.alta_v2.physics.utils.NpcMovementCalculator;
import com.alta_v2.physics.utils.TiledMapPhysicCalculator;
import com.alta_v2.rendering.actors.PersonView;
import com.badlogic.gdx.math.Vector2;
import lombok.Builder;
import lombok.Getter;

public class MoveNpcTask implements ResultTiledMapTask {

    private static final float EXECUTION_TIME_SECONDS = 0.2f;

    @Builder
    private static MoveNpcTask create(MovementDirection direction,
                                      ReservableActor npc,
                                      Vector2 targetPointLocal,
                                      AltitudeMap altitudeMap,
                                      ReservablePoint focusPointGlobal) {
        return new MoveNpcTask(direction, npc, targetPointLocal, altitudeMap, focusPointGlobal);
    }

    private final Vector2 targetPointGlobal;
    private final Vector2 targetPointLocal;
    private final float distanceLengthX;
    private final float distanceLengthY;
    private final ReservableActor npc;
    private final AltitudeMap altitudeMap;
    private final ReservablePoint focusPointGlobal;
    private final Tenant tenant = new Tenant("move-npc-task");

    private float currentExecutionTime = 0;

    @Getter
    private boolean isCompleted = false;

    @Getter
    private final BaseTaskResultObserver result = new BaseTaskResultObserver();

    private MoveNpcTask(MovementDirection direction,
                        ReservableActor npc,
                        Vector2 targetPointLocal,
                        AltitudeMap altitudeMap,
                        ReservablePoint focusPointGlobal) {
        this.npc = npc;
        this.altitudeMap = altitudeMap;
        this.focusPointGlobal = focusPointGlobal;

        this.targetPointLocal = targetPointLocal;
        this.targetPointGlobal = new Vector2(
                TiledMapPhysicCalculator.localToGlobal(targetPointLocal.x, altitudeMap.getTileWidth()),
                TiledMapPhysicCalculator.localToGlobal(targetPointLocal.y, altitudeMap.getTileHeight())
        );
        this.distanceLengthX = (this.targetPointLocal.x - this.npc.getLocalPoint().getX()) * altitudeMap.getTileWidth();
        this.distanceLengthY = (this.targetPointLocal.y - this.npc.getLocalPoint().getY()) * altitudeMap.getTileHeight();

        this.npc.reserve(this.tenant);
        this.npc.getLocalPoint().reserve(this.tenant);
        this.npc.getGlobalPoint().reserve(this.tenant);
        this.npc.getView().reserve(this.tenant).setValue(MovementDirection.getPersonView(direction), this.tenant);
    }

    @Override
    public void act(float delta) {
        if (this.isCompleted) {
            return;
        }

        this.currentExecutionTime += delta;
        if (this.currentExecutionTime >= EXECUTION_TIME_SECONDS) {
            this.npc.getGlobalPoint().setValue(this.targetPointGlobal, this.tenant);
        } else {
            // calculates the percentage of current time
            float currentPercentage = this.currentExecutionTime / EXECUTION_TIME_SECONDS * 100;

            // value to be used for one act.
            float reduceValueX = TiledMapPhysicCalculator.percentByValue(this.distanceLengthX, currentPercentage);
            float reduceValueY = TiledMapPhysicCalculator.percentByValue(this.distanceLengthY, currentPercentage);

            float globalX = NpcMovementCalculator.getPositionX(
                    npc.getLocalPoint().getX(), this.focusPointGlobal.getX(), this.altitudeMap.getTileWidth()
            );

            float globalY = NpcMovementCalculator.getPositionY(
                    npc.getLocalPoint().getY(), this.focusPointGlobal.getY(), this.altitudeMap.getTileHeight()
            );

            float currentX = this.npc.getGlobalPoint().getX() == this.targetPointGlobal.x ?
                    this.npc.getGlobalPoint().getX() : globalX + reduceValueX;

            float currentY = this.npc.getGlobalPoint().getY() == this.targetPointGlobal.y ?
                    this.npc.getGlobalPoint().getY() : globalY + reduceValueY;

            this.npc.getGlobalPoint().setValue(currentX, currentY, this.tenant);
        }

        this.postAct();
    }

    private void postAct() {
        if (this.npc.getGlobalPoint().getX() != this.targetPointGlobal.x ||
                this.npc.getGlobalPoint().getY() != this.targetPointGlobal.y) {
            return;
        }

        this.altitudeMap.setPointStatus(
                (int) this.npc.getLocalPoint().getX(),
                (int) this.npc.getLocalPoint().getY(),
                AltitudeMap.PointAvailability.FREE
        );

        this.altitudeMap.setPointStatus(
                (int) this.targetPointLocal.x, (int) this.targetPointLocal.y, AltitudeMap.PointAvailability.NPC
        );

        this.npc.getLocalPoint().setValue(this.targetPointLocal, this.tenant).release(this.tenant);
        this.npc.getGlobalPoint().release(this.tenant);
        this.npc.getView().release(this.tenant);
        this.npc.release(this.tenant);
        this.isCompleted = true;
        this.result.submitComplete();
    }

    @Override
    public void destroy() {
        this.result.destroy();
    }
}

File: core/src/com/alta_v2/physics/task/moveNpc/MoveNpcTask.kt
package com.alta_v2.physics.task.moveNpc

import com.alta_v2.physics.executionContext.Tenant
import com.alta_v2.physics.executionContext.altitude.AltitudeMap
import com.alta_v2.physics.executionContext.reserveData.ReservableActor
import com.alta_v2.physics.executionContext.reserveData.ReservablePoint
import com.alta_v2.physics.task.MovementDirection
import com.alta_v2.physics.task.ResultTiledMapTask
import com.alta_v2.physics.task.resultObserver.TaskResult
import com.alta_v2.physics.utils.NpcMovementCalculator
import com.alta_v2.physics.utils.TiledMapPhysicCalculator.Companion.localToGlobal
import com.alta_v2.physics.utils.TiledMapPhysicCalculator.Companion.percentByValue
import com.badlogic.gdx.math.Vector2

class MoveNpcTask(direction: MovementDirection,
                  private val npc: ReservableActor,
                  private val targetPointLocal: Vector2,
                  private val altitudeMap: AltitudeMap,
                  private val focusPointGlobal: ReservablePoint) : ResultTiledMapTask {

    companion object {
        private const val EXECUTION_TIME_SECONDS = 0.2f
    }

    private val targetPointGlobal = Vector2(
            localToGlobal(targetPointLocal.x, altitudeMap.tileWidth),
            localToGlobal(targetPointLocal.y, altitudeMap.tileHeight)
    )
    private val distanceLengthX: Float = (targetPointLocal.x - npc.localPoint.x) * altitudeMap.tileWidth
    private val distanceLengthY: Float = (targetPointLocal.y - npc.localPoint.y) * altitudeMap.tileHeight
    private val tenant = Tenant("move-npc-task")
    private var currentExecutionTime = 0f

    private var completed = false
    private val taskResult = TaskResult()

    init {
        altitudeMap.markAsBarrier(targetPointLocal.x.toInt(), targetPointLocal.y.toInt())
        npc.reserve(tenant)
        npc.localPoint.reserve(tenant)
        npc.globalPoint.reserve(tenant)
        npc.view.reserve(tenant).setValue(MovementDirection.getPersonView(direction), tenant)
    }

    override fun act(delta: Float) {
        if (isCompleted()) {
            return
        }
        currentExecutionTime += delta
        if (currentExecutionTime >= EXECUTION_TIME_SECONDS) {
            npc.globalPoint.setValue(targetPointGlobal, tenant)
        } else {
            // calculates the percentage of current time
            val currentPercentage = currentExecutionTime / EXECUTION_TIME_SECONDS * 100

            // value to be used for one act.
            val reduceValueX = percentByValue(distanceLengthX, currentPercentage)
            val reduceValueY = percentByValue(distanceLengthY, currentPercentage)
            val globalX = NpcMovementCalculator.getPositionX(
                    npc.localPoint.x, focusPointGlobal.x, altitudeMap.tileWidth
            )
            val globalY = NpcMovementCalculator.getPositionY(
                    npc.localPoint.y, focusPointGlobal.y, altitudeMap.tileHeight
            )
            val currentX = if (npc.globalPoint.x == targetPointGlobal.x) npc.globalPoint.x else globalX + reduceValueX
            val currentY = if (npc.globalPoint.y == targetPointGlobal.y) npc.globalPoint.y else globalY + reduceValueY
            npc.globalPoint.setValue(currentX, currentY, tenant)
        }
        postAct()
    }

    override fun isCompleted(): Boolean = completed

    override fun getResult(): TaskResult = taskResult

    private fun postAct() {
        if (npc.globalPoint.x != targetPointGlobal.x ||
                npc.globalPoint.y != targetPointGlobal.y) {
            return
        }
        altitudeMap.markAsFree(npc.localPoint.x.toInt(), npc.localPoint.y.toInt())
        altitudeMap.markAsObject(targetPointLocal.x.toInt(), targetPointLocal.y.toInt(), npc.id)
        npc.localPoint.setValue(targetPointLocal, tenant).release(tenant)
        npc.globalPoint.release(tenant)
        npc.view.release(tenant)
        npc.release(tenant)
        completed = true
        taskResult.complete(null)
    }

    override fun destroy() {
    }
}